File: java/com.suny.leetcode/Q94二叉树的中序遍历.java
import common.TreeNode;

import java.util.ArrayList;
import java.util.List;

/**
 * 94. 二叉树的中序遍历
 * 给定一个二叉树的根节点 root ，返回它的 中序 遍历。
 *
 * @author sunjianrong
 * @date 2021/6/30 下午9:50
 */
public class Q94二叉树的中序遍历 {


    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        inOrder(root, res);
        return res;
    }

    // 左 根 右
    private void inOrder(TreeNode root, List<Integer> res) {
        if (root == null) {
            return;
        }

        inOrder(root.left, res);
        res.add(root.val);
        inOrder(root.right, res);
    }
}

File: python/problems/Q94二叉树的中序遍历.py
# 94. 二叉树的中序遍历
# 给定一个二叉树的根节点 root ，返回 它的 中序 遍历 。

# 示例 1：
# 输入：root = [1,null,2,3]
# 输出：[1,3,2]
from typing import List, Optional

from python.problems.common.TreeNode import TreeNode


class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        res = []
        self.inOrders(res, root)
        return res

    def inOrders(self, res: List, treeNode: Optional[TreeNode]):
        if treeNode is None:
            return
        self.inOrders(res, treeNode.left)
        res.append(treeNode.val)
        self.inOrders(res, treeNode.right)
